### ISSUE TITLE
Returning unsized objects allowed in `Fn` traits but resulting values is "not a function"

### ISSUE THREAD

**Issue by hniksic** [rust-lang/rust#83189]
As explained in [the book](https://doc.rust-lang.org/book/ch19-05-advanced-functions-and-closures.html#returning-closures), a function cannot return an unsized object. So this function definition doesn't compile, as expected:

```rust
// error: "the return type of a function must have a statically known size"
fn returns_unsized() -> [u8] {
    todo!()
}
```

However, it is apparently possible to define an `Fn` trait that returns an unsized object. This compiles:

```rust
// compiles
type ReturnsUnsized = Box<dyn Fn() -> [u8]>;
fn bla(x: ReturnsUnsized) {
    todo!()
}
```

When you actually try to _use_ the type, compilation fails (which would be expected), but with an error message saying "call expression requires function" without pointing out the actual issue:

```rust
// error: "call expression requires function"
type ReturnsUnsized = Box<dyn Fn() -> [u8]>;
fn bla(x: ReturnsUnsized) {
    x();
}
```
```
error[E0618]: expected function, found `Box<(dyn Fn() -> [u8] + 'static)>`
 --> src/lib.rs:4:5
  |
3 | fn bla(x: ReturnsUnsized) {
  |        - `Box<(dyn Fn() -> [u8] + 'static)>` defined here
4 |     x();
  |     ^--
  |     |
  |     call expression requires function
```

I would expect the second snippet to also fail to compile. I would expect in both second and third snippet that the error message mention the underlying problem, which is returning an unsized object.

This issue was encountered by a [StackOverflow user](https://stackoverflow.com/questions/66646847/boxed-function-factory-higher-order-function-in-rust-results-in-call-expressi/66647090) who tried to declare a dynamic higher-order function as `Box<dyn Fn(f32, f32, f32, f32) -> (dyn Fn(f32) -> f32)>` (note unboxed return type) and call it. Making the outer closure return the inner closure boxed fixes the issue, but the error message was unhelpful in diagnosing the actual issue.

Reproducible in the [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3743e2b7226c0b88d54365a0b35d5ea9) with both stable, beta, and nightly compilers.

### Meta

`rustc --version`:
```
rustc 1.50.0
```

**Comment by jonas-schievink** [comment]
Hmm, I think this is a duplicate of some other bug, but can't find it now

**Comment by SNCPlay42** [comment]
#82633?

**Comment by hniksic** [comment]
#82633 appears to be about `fn() -> T` (note lower-case `fn`) being a valid type for unsized `T`, and implementing `FnOnce<(), Output = T>`. It also touches on soundness issues, especially in combination with `unsized_closures`.

This issue is about `Fn() -> T` being an apparently valid trait for unsized `T`, but being rejected (with an obscure error) when a type that implements it is actually called. It is more concerned with compiler diagnostics than with soundness.

It is quite possible that the underlying issue is the same and that they will be fixed together, but it's hard to be sure. Perhaps this one should be left open to ensure that the case described here is fixed?
